internal/services/compute/virtual_machine_scale_set.go

```diff
@@ -2222,3 +2224,87 @@ func flattenOrchestratedVirtualMachineScaleSetIdentity(input *identity.SystemAnd
+		if v, ok := d.GetOk("security_profile"); ok {
+			if values := securityprofile.FromBlock(v.([]interface{})); values != nil {
+				return *values
+			}
+		}
+
+		var values securityprofile.Values
+		if !features.FivePointOh() {
+			values.SecureBoot = pointer.To(d.Get("secure_boot_enabled").(bool))
+			values.VTPM = pointer.To(d.Get("vtpm_enabled").(bool))
+			if v, ok := d.GetOk("encryption_at_host_enabled"); ok {
+				values.HostEncryption = pointer.To(v.(bool))
+			}
+		}
+
+		return values
+	}()
+
+	encryptionType := virtualmachinescalesets.SecurityEncryptionTypes(securityEncryptionType)
+	if values.HostEncryption != nil && *values.HostEncryption {
+		if encryptionType == virtualmachinescalesets.SecurityEncryptionTypesDiskWithVMGuestState {
+			return nil, fmt.Errorf("`encryption_at_host_enabled` cannot be set to `true` when `os_disk.0.security_encryption_type` is set to `DiskWithVMGuestState`")
+		}
+	}
+
+	secureBootEnabled := pointer.From(values.SecureBoot)
+	vtpmEnabled := pointer.From(values.VTPM)
+	if encryptionType != "" {
+		if encryptionType == virtualmachinescalesets.SecurityEncryptionTypesDiskWithVMGuestState && !secureBootEnabled {
+			return nil, fmt.Errorf("`secure_boot_enabled` must be set to `true` when `os_disk.0.security_encryption_type` is set to `DiskWithVMGuestState`")
+		}
+		if !vtpmEnabled {
+			return nil, fmt.Errorf("`vtpm_enabled` must be set to `true` when `os_disk.0.security_encryption_type` is set")
+		}
+
+		const expected = string(virtualmachinescalesets.SecurityTypesConfidentialVM)
+		if values.SecurityType == nil {
+			values.SecurityType = pointer.To(expected)
+		} else if *values.SecurityType != expected {
+			return nil, fmt.Errorf("`security_profile.0.security_type` must be set to `%s` when `os_disk.0.security_encryption_type` is set", expected)
+		}
+	} else if secureBootEnabled || vtpmEnabled {
+		if values.SecurityType == nil {
+			values.SecurityType = pointer.To(string(virtualmachinescalesets.SecurityTypesTrustedLaunch))
+		} else {
+			switch v := *values.SecurityType; v {
+			case string(virtualmachinescalesets.SecurityTypesConfidentialVM):
+			case string(virtualmachinescalesets.SecurityTypesTrustedLaunch):
+			default:
+				return nil, fmt.Errorf("`security_profile.0.security_type` must not be set to `%s` when `secure_boot_enabled` or `vtpm_enabled` are set to true", v)
+			}
+		}
+	}
+
+	return &virtualmachinescalesets.SecurityProfile{
+		EncryptionAtHost: values.HostEncryption,
+		SecurityType:     (*virtualmachinescalesets.SecurityTypes)(values.SecurityType),
+		UefiSettings: &virtualmachinescalesets.UefiSettings{
+			SecureBootEnabled: values.SecureBoot,
+			VTpmEnabled:       values.VTPM,
+		},
+	}, nil
+}
+
+func flattenVirtualMachineScaleSetSecurityProfile(profile *virtualmachinescalesets.SecurityProfile) []interface{} {
+	if profile == nil {
+		return []interface{}{}
+	}
+
+	values := valuesFromVirtualMachineScaleSetSecurityProfile(profile)
+	return securityprofile.ToBlock(values)
+}
```

Comment:
could we change this to be in line with how we do other flattens in the provider?

Reply:
can you be more specific? I think this is the same comment as in batch_pool.go -- in other words, the idea here is to share code between the objects that have a common security_profile shape. Maybe you can tell me what you had in mind?